tests/test-coauthors-plus.php

```diff
@@ -679,6 +699,1434 @@ public function test_update_author_term_when_author_term_not_exist() {
+		$this->assertThat(
+			$guest_author_1,
+			$this->logicalNot(
+				$this->isInstanceOf( WP_User::class )
+			)
+		);
+
+		$post_id = $this->factory()->post->create(
+			array(
+				'post_author' => $this->author1->ID,
+				'post_status' => 'publish',
+				'post_type'   => 'post',
+			)
+		);
+
+		$query = new WP_Query(
+			array(
+				'p' => $post_id,
+			)
+		);
+
+		$this->assertEquals( 1, $query->found_posts );
+		$this->assertEquals( $this->author1->ID, $query->posts[0]->post_author );
+
+		$result = $this->_cap->add_coauthors(
+			$post_id,
+			array(
+				$guest_author_1->user_login,
+				$this->author3->user_login,
+			),
+			true
+		);
+
+		$this->assertTrue( $result );
+
+		$second_query = new WP_Query(
+			array(
+				'p' => $post_id,
+			)
+		);
+
+		$this->assertEquals( 1, $second_query->found_posts );
+		$this->assertEquals( $this->author1->ID, $second_query->posts[0]->post_author );
+
+		$post_author_terms = wp_get_post_terms( $post_id, $this->_cap->coauthor_taxonomy );
+
+		$this->assertTrue( is_array( $post_author_terms ) );
+		$this->assertCount( 3, $post_author_terms );
+
+		$author_slugs = array(
+			'cap-' . $this->author1->user_login,
+			'cap-' . $this->author3->user_login,
+			'cap-' . $guest_author_1->user_login,
+		);
+
+		foreach ( $post_author_terms as $term ) {
+			$this->assertInstanceOf( WP_Term::class, $term );
+			$this->assertContains( $term->slug, $author_slugs );
+		}
+	}
+
+	/**
+	 * This is where we test many moving parts of the CoAuthorsPlugin all at once. We are creating a guest author from a
+	 * WP_User, and then assigning that guest author to a post. Since the guest author is linked to a WP_User, the
+	 * function CoAuthors_Plus::get_coauthor_by() should return a guest author object along with meta data
+	 * indicating that the object is linked to a WP_User. The wp_posts.post_author column should change,
+	 * and the response from CoAuthors_Plus::add_coauthors() should be true.
+	 * @return void
+	 */
+	public function test_assign_post_authors_from_coauthors_who_are_linked() {
+		$this->_cap->guest_authors->create_guest_author_from_user_id( $this->author2->ID );
+		$this->_cap->guest_authors->create_guest_author_from_user_id( $this->author3->ID );
+
+		$linked_author_1 = $this->_cap->get_coauthor_by( 'id', $this->author2->ID );
+		$this->assertIsObject( $linked_author_1 );
+		$this->assertThat(
+			$linked_author_1,
+			$this->logicalNot(
+				$this->isInstanceOf( WP_User::class )
+			)
+		);
+		$this->assertTrue( property_exists( $linked_author_1, 'type' ) );
+		$this->assertEquals( 'guest-author', $linked_author_1->type );
+		$this->assertTrue( property_exists( $linked_author_1, 'is_wp_user' ) );
+		$this->assertTrue( $linked_author_1->is_wp_user );
+		$this->assertTrue( property_exists( $linked_author_1, 'wp_user' ) );
+		$this->assertInstanceOf( WP_User::class, $linked_author_1->wp_user );
+		$this->assertEquals( $this->author2->ID, $linked_author_1->wp_user->ID );
```

Comment:
It strongly feels like these assertions should go into a test specifically for `get_coauthor_by()` - so that this test method can just call the happy path of `get_coauthor_by()` and focus on what happens when the guest author is added with `add_coauthors()`. i.e. the whole test should be something like:

```php
// Arrange:
// Create linked guest author from User 1.
// Create linked guest author from User 2.
// Create post with Editor 1 as the author.

// Act:
// Add both guest authors to the post.

// Assert:
// Assert the post author is User 1.
```

That should be it. All the other assertions should be handled (happy and sad paths) in other tests. The only thing this test should be doing is adding multiple ordered linked tests and see if the post_author field was updated.